### PR TITLE
Methods that returns live node collection fixed (closes #1096)

### LINKED ISSUE #1096: Wrong proxing methods that returns live node collection
Reproduced in playrground. Relates to https://github.com/DevExpress/testcafe/issues/1442

### PR DESCRIPTION
**Pull request by georgiy-abbasov**
\cc @churkin, @miherlosev

### CODE REVIEW

**Comment by georgiy-abbasov** [comment]
FPR

File: src/client/sandbox/node/node-list-wrapper.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        this.filteredNodeList = this.filterListFn(this.nodeList, this.filterItemFn);
+
+        var wrapper = this;
+        var i;
+
+        for (i = 0; i < this.filteredNodeList.length; i++)
+            Object.defineProperty(wrapper, i.toString(), { configurable: true, value: wrapper.filteredNodeList[i] });
+
+        for (i = this.filteredNodeList.length; i < this.prevNodeListLength; i++)
+            Object.defineProperty(wrapper, i.toString(), { configurable: true, value: void 0 });
```
Comment (miherlosev) [inline src/client/sandbox/node/node-list-wrapper.js]: Can we just delete property from wrapper?
Like this 
```js
for (i = this.filteredNodeList.length; i < this.prevNodeListLength; i++)
         delete  wrapper[i];
```

File: src/client/sandbox/shadow-ui.js [hunk 2]
```diff
@@ -140,15 +148,15 @@ export default class ShadowUI extends SandboxBase {
         };
 
         docProto.getElementsByClassName = function (...args) {
-            return ShadowUI._filterNodeList(nativeMethods.getElementsByClassName.apply(this, args));
+            return ShadowUI._filterNodeList(nativeMethods.getElementsByClassName.apply(this, args), true);
```
Comment (miherlosev) [inline src/client/sandbox/shadow-ui.js]: It's difficult to read.
Let rewrite as
```js
return ShadowUI._filterNodeList({ 
    nodeList:         nativeMethods.getElementsByClassName.apply(this, args),     
    isLiveCollection: true
});
```

File: test/client/fixtures/sandbox/shadow-ui-test.js [hunk 3]
```diff
@@ -733,3 +733,93 @@ if (document.implementation && document.implementation.createHTMLDocument) {
         strictEqual(htmlDoc.getElementsByTagName('body')[0], nativeMethods.getElementsByTagName.call(htmlDoc, 'body')[0]);
     });
 }
+
+module('"getElements" methods should return live node collections (GH-1096)');
+
+test('getElementsByClassName', function () {
```
Comment (miherlosev) [inline test/client/fixtures/sandbox/shadow-ui-test.js]: For all appropriate methods add `instanceof` check
For example:
```js
ok(document.getElementsByTagName(...) instanceof HTMLCollection)
```

File: src/client/sandbox/node/node-list-wrapper.js (hunk 1 above)
Comment (georgiy-abbasov) [inline src/client/sandbox/node/node-list-wrapper.js]: I tried, it didn't work

Comment (miherlosev) [inline src/client/sandbox/node/node-list-wrapper.js]: I found that `delete` operator is very slowly - https://jsperf.com/delete-vs-undefined-vs-null/16

File: src/client/sandbox/node/node-list-wrapper.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
+export default class NodeListWrapper {
+    constructor (nodeList, filterListFn, filterItemFn) {
+        this.nodeList           = nodeList;
```
Comment (miherlosev) [inline src/client/sandbox/node/node-list-wrapper.js]: It is private variables.
Let's start them with `_` and hide from enumeration

**Comment by georgiy-abbasov** [comment]
I found one more error, which may occur during test execution. It's not relate to this PR, but it should be fixed here, i assume.

**Comment by georgiy-abbasov** [comment]
FPR

File: src/client/sandbox/node/node-list-wrapper.js [hunk 5]
```diff
@@ -0,0 +1,85 @@
+export default class NodeListWrapper {
+    constructor (nodeList, filterListFn, filterItemFn) {
+        Object.defineProperty(this, 'item', {
+            enumerable:   true,
+            configurable: false,
```
Comment (miherlosev) [inline src/client/sandbox/node/node-list-wrapper.js]: Default value for `configurable`, `emumerable`, `writable` properties is `false`.
So, you can just remove `configurable: false`, `enumerable:   false`;

File: src/client/sandbox/shadow-ui.js [hunk 6]
```diff
@@ -56,8 +57,15 @@ export default class ShadowUI extends SandboxBase {
...
 
-    static _filterNodeList (nodeList) {
-        return ShadowUI._filterList(nodeList, item => ShadowUI._filterElement(item));
+    static _filterNodeList ({ nodeList, isLiveCollection }) {
+        var filterItemFn = item => ShadowUI._filterElement(item);
+
+        if (!isLiveCollection)
+            return ShadowUI._filterList(nodeList, filterItemFn);
+
+        NodeListWrapper.prototype = nodeList;
```
Comment (miherlosev) [inline src/client/sandbox/shadow-ui.js]: Move to line 62 and add `instanceof` check for `querySelector` function

File: test/client/fixtures/sandbox/shadow-ui-test.js [hunk 7]
```diff
@@ -733,3 +734,102 @@ if (document.implementation && document.implementation.createHTMLDocument) {
...
+    div3.className = testClassName;
+
+    testDiv.appendChild(div1);
+    testDiv.appendChild(div2);
+    testDiv.appendChild(div3);
+
+    shadowUI.addClass(div3, 'el');
+
+    var elements         = document.getElementsByClassName(testClassName);
+    var expectedInstance = browserUtils.isSafari && browserUtils.version < 10 ? NodeList : HTMLCollection;
```
Comment (miherlosev) [inline test/client/fixtures/sandbox/shadow-ui-test.js]: Rename to `expectedType`

File: src/client/sandbox/code-instrumentation/properties/index.js [hunk 8]
```diff
@@ -648,7 +648,7 @@ export default class PropertyAccessorsInstrumentation extends SandboxBase {
 
             scripts: {
                 condition: domUtils.isDocument,
-                get:       doc => ShadowUI._filterNodeList(doc.scripts),
+                get:       doc => ShadowUI._filterNodeList({ nodeList: doc.scripts, isLiveCollection: true }),
```
Comment (miherlosev) [inline src/client/sandbox/code-instrumentation/properties/index.js]: Add test for this change.

**Comment by georgiy-abbasov** [comment]
@testcafe-build-bot \retest

**Comment by georgiy-abbasov** [comment]
@testcafe-build-bot \retest

**Comment by georgiy-abbasov** [comment]
@testcafe-build-bot \retest

File: src/client/sandbox/shadow-ui.js [hunk 9]
```diff
@@ -56,8 +57,15 @@ export default class ShadowUI extends SandboxBase {
         return filteredList.length === nlLength ? list : filteredList;
     }
 
-    static _filterNodeList (nodeList) {
-        return ShadowUI._filterList(nodeList, item => ShadowUI._filterElement(item));
+    static _filterNodeList ({ nodeList, isLiveCollection }) {
```
Comment (churkin) [inline src/client/sandbox/shadow-ui.js]: May be `static _filterNodeList (nodeList [, isLiveCollection])`?

Comment (miherlosev) [inline src/client/sandbox/shadow-ui.js]: Object initializer was added to improve readability.
compare:
```js
return ShadowUI._filterNodeList({
                nodeList:         nativeMethods.getElementsByClassName.apply(this, args),
                isLiveCollection: true
});
```

versus
```js
return ShadowUI._filterNodeList(nativeMethods.getElementsByClassName.apply(this, args), true);
```

Comment (churkin) [inline src/client/sandbox/shadow-ui.js]: or

```javascript
var nodeList = nativeMethods.getElementsByClassName.apply(this, args);

return ShadowUI._filterNodeList(nodeList, true);
```

Comment (miherlosev) [inline src/client/sandbox/shadow-ui.js]: Ok.
Then you are reading `return ShadowUI._filterNodeList(nodeList, true);` you cannot understand that mean `true` parameter without navigate to definition.

Comment (AlexanderMoskovkin) [inline src/client/sandbox/shadow-ui.js]: I don't see so big problem here, but I vote for `ShadowUI._filterNodeList(nodeList, isLiveCollection)` to be consistent with other filter methods (filterElement, filterList) where the first argument is an element/list.

Comment (georgiy-abbasov) [inline src/client/sandbox/shadow-ui.js]: Ok

**Comment by georgiy-abbasov** [comment]
FPR

**Pull request merged**